classes/task/merge_user_accounts.php

```diff
@@ -0,0 +1,184 @@
+                                            ['id' => $data->mergerequestid]);
+        if (!$mergerequest) {
+            // Merge request was removed before running this task.
+            return;
+        }
+        $lookatcriteriaforusers = $this->verify_users_to_keep_and_remove($mergerequest);
+        $mergerequest = $this->merge($record, $maxattempts, merge_request::TRIED_WITH_ERROR);
+        if ($mergerequestresult->status == merge_request::COMPLETED_WITH_SUCCESS) {
+            /* We run the merge request AGAIN because the user may be interacting with Moodle
+            * while merge request is being processed, so that NO ALL records are correctly migrated
+            * into the user to keep. It will ensure ALL records are correctly migrated into the user to keep. */
+            $mergerequestresult = $this->merge($mergerequestresult, $maxattempts, merge_request::COMPLETED_WITH_ERRORS);
+        }
+        if ($mergerequestresult->status != merge_request::COMPLETED_WITH_ERRORS &&
+            $mergerequestresult->status != merge_request::COMPLETED_WITH_SUCCESS) {
+            // Throwing exception will ensure this adhoc task is re-queued until $maxretries is reached.
+            throw new moodle_exception(get_string('failedmergerequest', 'tool_mergeusers'));
+        }
+    }
+    /**
+     * Function for merging two users.
+     */
+    private function merge(object $record, int $maxattempts, int $statusforerror) {
+        $retries = $record->retries + 1;
+        // Update retries.
+        $this->update_retries_in_table($record->id, $retries);
+        $logs = $record->log;
+        $mut = new MergeUserTool();
+        list($success, $log, $logid) = $mut->merge($record->keepuserid, $record->removeuserid);
+        if ($success) {
+            $status = merge_request::COMPLETED_WITH_SUCCESS;
+        } else {
+            if ($retries >= $maxattempts) {
+                $status = merge_request::COMPLETED_WITH_ERRORS;
+                // Send a notification to administrator ?
+            } else {
+                $status = $statusforerror;
+            }
+        }
+        $logs[$retries] = $log;
+        $this->update_status_and_log_in_table($record->id,
+                                            $status,
+                                            $logs);
+        $record->log = $logs;
+        $record->status = $status;
+        return $record;
+    }
+    /**
+     * Function for updating the status of the merging request.
+     */
+    private function update_status_and_log_in_table(int $idrecord,
+                                                    int $status,
+                                                    array $log) {
+        global $DB;
+        $DB->update_record(
+            merge_request::TABLE_MERGE_REQUEST,
+            (object)[
+                'id' => $idrecord,
+                'status' => $status,
+                'timecompleted' >= time(),
+                'log' => json_encode($log),
```

Comment:
As it is done in the merge_request class, I would rather prefer to hide the logic of log structure inside the merge_request class. So, if we were using an instance of merge_request would be better and simpler from the client side using all these fields. Instead of passing the 3 params, just pass the merge_request instance. If necessary, add some method to return logs without decoding (without using json_decode()), so log field could be obtained for storing.

Reply:
Proceed as you consider. If you like to leave it as it is, go on.

However, I think that `timemodified` should be updated any time any record is updated. However, thinking about it a bit deeper, `timecompleted` should be only updated when `$status` is considered as some of the completion status, like `merge_request::COMPLETED_WITH_SUCCESS` or `merge_request::COMPLETED_WITH_ERROR`. Thanks.